torch/csrc/distributed/rpc/python_rpc_handler.cpp

```diff
@@ -16,6 +16,19 @@ py::object getFunction(const py::object& module, const char* name) {
   return fn;
 }
 
+// if cachedFn is not none, assign it to local function; otherwise, import
+// function from the scratch and assign it to local function;
+py::object getLocalFunction(const py::object& cachedFn, const char* name) {
```

Comment:
why do we need this given that the three py functions are already loaded in the constructor?

Reply:
because the three py functions will be cleaned up in Rpc agent join(), but they could be called after Rpc agent join() in rare cases. 

for most cases, we do not need to load them again, just copy cachedFn to localFn, this should be cheap as it is just copying pointer. 

Does it make sense?

Reply:
> but they could be called after Rpc agent join() in rare cases.

when could this happen? And if we load them again, do we need to cleanup them again?

Reply:
e.g.  loadPythonUDFResult() is called in future.wait(), which could be called after Rpc agent join(). 

it does not need to clean up them again, as it is loaded to a local variable intentionally for this rare case

Reply:
I see. Let's get this PR in to fix master, and let's discuss whether we need to support the above case. I feel that all `future.wait()` in user code should be resolved before calling `join`, and if we detect termination in `join()`, `RequestCallbackImpl` won't hit this case either (?).

Reply:
@mrshenli yes, RequestCallbackImpl will not hit this case. but I still add it for runPythonUDF() function for safety reason, because in theory functions in PythonRpcHandler could be called anytime.